### PR TITLE
Allow i2s microphone bits per sample to be configured

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This brings a significant increase in audio quality for some microphones and allows others to work at all.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2943

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/i2s_audio/microphone/i2s_audio_microphone.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ void I2SAudioMicrophone::start_() {
   i2s_driver_config_t config = {
       .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_RX),
       .sample_rate = 16000,
-      .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
+      .bits_per_sample = I2S_BITS_PER_SAMPLE_32BIT,
```
Comment (sehraf) [inline esphome/components/i2s_audio/microphone/i2s_audio_microphone.cpp]: Doesn't this break external ADCs, which expect 16 bits per sample and (currently) can't be reconfigured?

Comment (jesserockz) [reply esphome/components/i2s_audio/microphone/i2s_audio_microphone.cpp]: Potentially maybe...
I guess I can make it configurable and dynamically scale the data if you have it set to not 16-bit which was on my mind to do.
Thanks for the push =)

**Comment by jesserockz** [comment]
Thanks for the feedback @sehraf, I have now made the input bits configurable and it will scale down to 16-bit before passing to other components like voice_assistant.

**Comment by kroimon** [comment]
While I like the idea very much in general, I wonder if we should really add the new `bits_per_sample` option to the microphone or if it wouldn't be better to add it to the base `i2s` component.

The current I2S media player aside, the sample rate and bit depth are inherent parameters of the bus itself and if we want to move to a more flexible implementation allowing full-duplex use of microphone and speaker, this might be a breaking change later.
For reference, [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/i2s.html#full-duplex) is the latest esp-idf i2s driver documentation for full-duplex mode for the ESP32-S3.

**Comment by kroimon** [comment]
Also, having these settings go through the main `i2c` component as a hub would allow components interfacing external ADC/DAC audio codecs to get the current configuration from there to be sent to the codecs as those require having a matching configuration.

**Comment by jesserockz** [comment]
> While I like the idea very much in general, I wonder if we should really add the new bits_per_sample option to the microphone or if it wouldn't be better to add it to the base i2s component.

ESPHome does not support duplex (And I dont see a need for it in the near future) and each component (microphone/media_player) reconfigure the i2s bus with their own individual config. This allows mics and speakers of different bits/channels etc to be able to work "together" on the same board.

The only reason for the common i2s_audio component with the clock pins is so that multiple components do not try to (re)configure the bus at the same time and it holds a lock per component to prevent this.

**Pull request merged**